compiler/rustc_typeck/src/check/cast.rs

```diff
@@ -551,20 +560,54 @@ impl<'a, 'tcx> CastCheck<'tcx> {
         });
     }
 
+    /// invald-ptr-to-int-cast lint: when casting a function pointer to a numeric type it only
+    /// makes sense to cast to `usize`. Other cases are often bugs. See #81686 as an example.
+    fn invalid_ptr_to_int_cast_lint(&self, fcx: &FnCtxt<'a, 'tcx>) {
+        if self.expr_ty.is_fn() || self.expr_ty.is_raw_ptr() {
+            let is_integral_cast = matches!(self.cast_ty.kind(), ty::Int(_) | ty::Uint(_));
+            if is_integral_cast && !matches!(self.cast_ty.kind(), ty::Uint(ty::UintTy::Usize)) {
+                fcx.tcx.struct_span_lint_hir(
+                    lint::builtin::INVALID_PTR_TO_INT_CAST,
+                    self.full_expr.hir_id,
+                    self.full_expr.span,
+                    |lint| {
+                        let msg = format!("casting pointer to `{}`", self.cast_ty);
+                        let mut lint = lint.build(&msg);
+                        lint.help("pointers should only be cast to `usize`");
+
+                        if let Ok(e_snippet) =
+                            fcx.tcx.sess.source_map().span_to_snippet(self.expr.span)
+                        {
+                            lint.span_suggestion(
+                                self.full_expr.span,
+                                &format!("to cast to `{}`, cast to `usize` first", self.cast_ty),
+                                format!("{} as usize as {}", e_snippet, self.cast_ty),
+                                Applicability::MachineApplicable,
```

Comment:
```suggestion
                                Applicability::MaybeIncorrect,
```
This presence of this lint means that we expect the user not to have intended to write the cast, so likely the user will want to rewrite their code, rather than ignore the lint by casting via `usize`.